### PR TITLE
Event Api Complete

### PR DESCRIPTION
**Pull request by jakesboy2**
This is not final!
Needs to be bug tested, and cleaned up before merging
I'm opening the pull request so we can catch obvious errors in syntax or in logic
Just need an extra eye on this. Approve when you have looked over it, and I will clean and refactor messy areas and fix bugs before merging in.

### CODE REVIEW

File: src/Helpers/EventHelper.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        let requestingFlag: AttackFlagMemory | undefined;
+
+        // Find the one that requested this creep and handle it accordingly
+        // !TODO Before Merging pullrequest and finish bug fixes
+        // Move this into a function, no reason the function should be this crowded with nonsense
+        // Also, brak plz, help me find a way to standarize this so we don't have to edit some cryptic function
+        // every time we want to add a military role
+        // If not removing that entirely, at least help make it easy to edit please lol
+        for (const attackFlag of attackRoomFlags) {
+
```
Comment (jakesboy2) [inline src/Helpers/EventHelper.ts]: Need to add a case here to make sure the creeps target room is the same as the attack flags room name, so we don't mix up squads

File: src/Helpers/EventHelper.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            (structure: Structure) =>
+                structure.structureType !== STRUCTURE_KEEPER_LAIR
+                && structure.structureType !== STRUCTURE_CONTROLLER &&
+                (structure.memory.processed === false || structure.memory.processed === undefined),
+            true
+        );
+
+
+        // Create an event for each unprocesseed structure, then set it as processed
+        for (const key in structures) {
```
Comment (jakesboy2) [inline src/Helpers/EventHelper.ts]: Make sure i test here to ensure this error is actually fixed, its game breaking

File: src/Helpers/RoomHelper.ts [hunk 3]
```diff
@@ -199,9 +199,9 @@ export default class RoomHelper {
             throw new UserException(
                 "Tried to getUpgraderLink of a room with no controller",
                 "Get Upgrader Link was called for room [" +
-                    room.name +
```
Comment (jakesboy2) [inline src/Helpers/RoomHelper.ts]: My linter is superior to yours

File: src/Helpers/RoomHelper.ts [hunk 4]
```diff
@@ -452,12 +452,68 @@ export default class RoomHelper {
...
+            return false;
+        }
+        if (!Memory.rooms[hostRoomName]) {
+            return false;
+        }
+        if (!Game.rooms[hostRoomName]) {
+            return false;
+        }
+
+        const remoteRooms: RemoteRoomMemory[] = MemoryApi.getRemoteRooms(Game.rooms[hostRoomName]);
```
Comment (jakesboy2) [inline src/Helpers/RoomHelper.ts]: No reason that getRemoteRooms and related functions needs vision in the room, on next major refactor we can fix this to just require room.name as we have slowly been doing

File: src/utils/config.ts [hunk 5]
```diff
@@ -124,8 +131,10 @@ export const SPAWN_MANAGER_BUCKET_LIMIT = 50;
...
 export const ROOM_MANAGER_BUCKET_LIMIT = 500;
 export const MEMORY_MANAGER_BUCKET_LIMIT = 1;
+export const EVENT_MANAGER_BUCKET_LIMIT = 5000;
+export const ROOM_OVERLAY_BUCKET_LIMIT = 7000;
 
 /**
  * List of allies
  */
-export const ALLY_LIST: string[] = ["jakesboy2", "Uhmbrock"];
+export const ALLY_LIST: string[] = ["jakesboy2", "Uhmbrock", "Apexia", "SquishProxy"];
```
Comment (jakesboy2) [inline src/utils/config.ts]: Need to remove this lol

File: src/utils/kungfu.ts [hunk 6]
```diff
@@ -0,0 +1,294 @@
+
```
Comment (jakesboy2) [inline src/utils/kungfu.ts]: Standard library file, will update manually as i add to the actual library, but would be nice to find a way to hook it to git automatically or add into node modules

File: src/Helpers/EventHelper.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        // Solved, defined the property for it. should be fine now
+
+        // Get all structures in the room that have not yet been processed
+        const structures: Structure[] = MemoryApi.getStructures(
+            room.name,
+            (structure: Structure) =>
+                structure.structureType !== STRUCTURE_KEEPER_LAIR
+                && structure.structureType !== STRUCTURE_CONTROLLER &&
+                (structure.memory.processed === false || structure.memory.processed === undefined),
+            true
```
Comment (UhmBrock) [inline src/Helpers/EventHelper.ts]: I would prefer if we don't invalidate the cache every tick.. Would it be possible to limit this?

Comment (jakesboy2) [reply src/Helpers/EventHelper.ts]: currently not possible to do this and have the bugs this fixes, still be fixed (requires finding things on the same tick, could only trigger invalidation of cache on a certain trigger that would indicate something like this happened (ie room events has a creep build action in it or something)

**Pull request merged**